lib/Doctrine/Common/Reflection/TypedNoDefaultReflectionProperty.php

```diff
@@ -20,4 +20,29 @@ public function getValue($object = null)
     {
         return $object !== null && $this->isInitialized($object) ? parent::getValue($object) : null;
     }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Works around the problem with setting typed no default properties to
+     * NULL which is not supported, instead unset() to uninitialize.
+     *
+     * @link https://github.com/doctrine/orm/issues/7999
+     */
+    public function setValue($object, $value = null)
+    {
+        if ($value === null) {
+            $propertyName = $this->getName();
+
+            $unsetter = function () use ($propertyName) {
+                unset($this->$propertyName);
+            };
+            $unsetter = $unsetter->bindTo($object, $this->getDeclaringClass()->getName());
+            $unsetter();
+
+            return;
+        }
+
+        parent::setValue($object, $value);
```

Comment:
Is it worth to keep a property with the original closure? Doing that is safe since `Closure#bindTo()` is a immutable operation. It would avoid creating maaaaany unnecessary closures in runtime but we should benchmark it to see the impact (memory/processing time.)

That would simplify this quite a lot, considering the unsetter is created at construction time:

```suggestion
        if ($value === null) {
            $this->unsetter->bindTo($object, $this->getDeclaringClass()->getName())();
            
            return;
        }
        
        parent::setValue($object, $value);
```

Reply:
We could also save the result of `$this->getDeclaringClass()->getName()` to reduce call count to these methods, btw.

Reply:
I wouldn't optimize this yet, because i don't see this happening often. It is needed for deleting entities for sure, when ORM sets the id to null. But why would users have a null value for a typed, non-null property themselves? That be a weird usage I don't recommend anyways